src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/util/SSLConnectionTestUtil.java

```diff
@@ -0,0 +1,189 @@
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.security.ssl.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Utility class to test if the server supports SSL connections.
+ * SSL Check will be done by sending an ES Ping to see if server is replying to pings.
+ * Following that a custom client hello message will be sent to the server, if the server
+ * side has OpenDistroPortUnificationHandler it will reply with server hello message.
+ */
+public class SSLConnectionTestUtil {
+
+    private static final Logger logger = LogManager.getLogger(SSLConnectionTestUtil.class);
+    public static final byte[] ES_PING_MSG = new byte[]{(byte) 'E', (byte) 'S', (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF};
+    public static final String DUAL_MODE_CLIENT_HELLO_MSG = "DUALCM";
+    public static final String DUAL_MODE_SERVER_HELLO_MSG = "DUALSM";
+    private static final int SOCKET_TIMEOUT_MILLIS = 10 * 1000;
+    private boolean esPingReplyReceived;
+    private boolean dualSSLProbeReplyReceived;
+    private final String host;
+    private final int port;
+    private Socket overriddenSocket = null;
+    private OutputStreamWriter testOutputStreamWriter = null;
+    private InputStreamReader testInputStreamReader = null;
+
+    public SSLConnectionTestUtil(final String host, final int port) {
+        this.host = host;
+        this.port = port;
+        esPingReplyReceived = false;
+        dualSSLProbeReplyReceived = false;
+    }
+
+    @VisibleForTesting
+    protected SSLConnectionTestUtil(final String host, final int port, final Socket overriddenSocket, final OutputStreamWriter testOutputStreamWriter,
+        final InputStreamReader testInputStreamReader) {
+        this.overriddenSocket = overriddenSocket;
+        this.testOutputStreamWriter = testOutputStreamWriter;
+        this.testInputStreamReader = testInputStreamReader;
+
+        this.host = host;
+        this.port = port;
+        esPingReplyReceived = false;
+        dualSSLProbeReplyReceived = false;
+    }
+
+    /**
+     * Test connection to server by performing the below steps:
+     * - Send Client Hello to check if the server replies with Server Hello which indicates that Server understands SSL
+     * - Send ES Ping to check if the server replies to the ES Ping message
+     *
+     * @return SSLConnectionTestResult i.e. ES_PING_FAILED or SSL_NOT_AVAILABLE or SSL_AVAILABLE
+     */
+    public SSLConnectionTestResult testConnection() {
+        if (sendDualSSLClientHello()) {
+            return SSLConnectionTestResult.SSL_AVAILABLE;
```

Comment:
this mean other node is able to communicate in DualMode (and hence SSL too)

But what will happen if other node is only listening in SSL mode  (and not in Dual Mode) ? 
 This check (sendDualSSLClientHello() ) will fail..  and also SendESPing() will fail .. (but actually SSL is still available so node should use SSL to communicate)

Reply:
This may happen in ODFE installation when new node with enabled dual SSL mode is added to an existing cluster with prior version of security plugin installed. If it is not supported, there should be clear error message, that dual SSL mode is not supported, not just network timeout.

Reply:
I am still not getting why would this happen, am i missing anything ? why would customer enable dual mode when existing cluster has SSL plugin and scan speak SSL already ? Ideally they must be following doc and enabling dual mode when migrating existing cluster to add security plugin with SSL enabled. It just wont work as customer has configured it wrongly. Correct us if we are wrong.

Reply:
It may be enabled by accident in which case ES should terminate with a reasonable error message and not be stuck  waiting for a response that will never come.

Additionally, it looks that the handshake may require establishing 2 **extra** socket connections (one long lived connection and 2 short lived socket connections). Why one connection is not sufficient?

Reply:
I have updated the error message to be more clearer. 

Regarding the 2 extra sockets, we will be opening only 1 extra socket in the best case, i.e. if transport server supports dual mode. And 2 sockets when transport server does not support SSL. 
The short lived connections are required to probe the transport server to check if it supports SSL connection.
Also, these extra sockets are opened only when dual mode is enabled (i.e. during EIT migration phase), in all other scenarios this part of code is never hit.

The following scenarios are possible:
- Dual mode is enabled; Transport client and transport server both support dual mode:
    - Transport client opens a socket to perform a custom handshake with the transport server
    - Transport server responds to custom handshake and closes the socket (since this is a plain text socket and cant be used any more)
    - Transport client now knows that the server supports SSL and proceeds with SSL connection (full SSL Handshake will take place)
- Dual mode is enabled; Transport client supports dual mode, transport server does not support dual mode and dual mode is enabled:
    - Transport client opens a socket to perform custom handshake with transport server. 
    - Transport server does not recognize this message and closes the connection. 
    - Transport client see that connection is closed and opens another socket to send ES Ping
    - Transport server responds to this ES ping. 
    - Transport client will proceed with plain text connection with this transport server.
- Dual mode is disabled:
    - No extra sockets are opened, SSL connections are setup by default.